Neos.ContentRepository/Classes/Domain/Projection/Content/TraversableNodeInterface.php

```diff
@@ -0,0 +1,121 @@
+<?php
+declare(strict_types=1);
+
+namespace Neos\ContentRepository\Domain\Projection\Content;
+
+/*
+ * This file is part of the Neos.ContentRepository package.
+ *
+ * (c) Contributors of the Neos Project - www.neos.io
+ *
+ * This package is Open Source Software. For the full copyright and license
+ * information, please view the LICENSE file which was distributed with this
+ * source code.
+ */
+
+use Neos\ContentRepository\Domain\ValueObject\NodeName;
+use Neos\ContentRepository\Domain\ValueObject\NodePath;
+use Neos\ContentRepository\Domain\ValueObject\NodeTypeConstraints;
+use Neos\ContentRepository\Exception\NodeException;
+use Neos\EventSourcedContentRepository\Domain\Projection\Content\ContentSubgraphInterface;
+use Neos\EventSourcedContentRepository\Domain\ValueObject\PropertyName;
+
+/**
+ * This is a NEW interface, introduced in Neos 4.3; and it will become the main interface
+ * with Neos 5.0 to the CR.
+ *
+ * The main convenience Event-Sourced NodeInterface used for READING; containing
+ * data accessors and traversal methods.
+ *
+ * All methods which are called `get*()` contain only information local to a node,
+ * so they can be accessed really quickly without any external lookup.
+ *
+ * All methods which are called `find*()` may involve some database querying to
+ * fetch their information.
+ *
+ * The TraversableNodeInterface is *immutable*, meaning its contents never change after creation.
+ * It is *only used for reading*.
+ *
+ * Starting with version 5.0 (when backed by the Event Sourced CR), it is
+ * *completely detached from storage*; so it will not auto-update after a property changed in
+ * storage.
+ */
+interface TraversableNodeInterface extends NodeInterface
+{
+    public function getSubgraph(): ContentSubgraphInterface;
```

Comment:
Can we please skip this until we've established that we really really need it? I'd hate to see the TraversableNodeInterface mutate from a convenience wrapper to a God Object unless absolutely necessary. The requested subgraph should be readily available anywhere.

Reply:
Hey,

@nezaniel We currently need it at various places in the core; as the requested subgraph is not available everywhere. Of course one could start fetching the subgraph from the graph using the DSP and ContentStreamIdentifier, but you cannot drop both at the same time from the interface :) 

I'm still against dropping this, as this makes the codebase at various places a lot harder where we need exactly the combination of a (Traversable)Node and a subgraph for further queries.

When we have a PR ready which can live without this, we can drop this from the interface afterwards as well, but I wouldn't want delaying getting this PR for the core for this.

All the best,
Sebastian

Reply:
Hey,

By core do you mean the legacy core or the new one? In legacy case I'd say keep this and mark it as deprecated, if it's the new one then let's please add it later. Adding something to an interface is massively more easy than removing it once it is live. I think we have some fundamental achitectural issues to discuss here since I am convinced that we neither need a NodeInterface's requested DSP nor a TraversableNodeInterface's subgraph via API and would like to see some really convincing examples first. I'll have a look into this asap.

Best,
Bernhard

Reply:
@nezaniel let's speak about code, I am oppositely convinced it definitely makes sense to have at least one of them available :D ;) Of course as soon as you have a working PR, I'm open to be convinced otherwise ;)

Reply:
This is a difficult one, but I have to say I'm also more on the "conservative" side of things in this matter:
The legacy Node [does not implement this method](https://github.com/neos/neos-development-collection/blob/event-sourced-patch/Neos.ContentRepository/Classes/Domain/Model/Node.php#L114-L117) and throws an exception instead and the new implementation is called only in [RenderContentOutOfBand::renderContent()](https://github.com/neos/contentrepository-development-collection/blob/master/Neos.EventSourcedNeosAdjustments/Classes/Ui/Domain/Model/Feedback/Operations/RenderContentOutOfBand.php#L221) AFAIS..
So IMO we could defer the decision by removing the method from the interface and adding an `instanceof` check for now

Reply:
nope @bwaidelich , you also need to search the codbase in Fusion for ".subgraph" to find all occurences where this is used.

All the best,
Sebastian

Reply:
> you also need to search the codbase in Fusion for ".subgraph"

Ah, good point..!
But.. I searched the project for "subgraph" in Fusion files and can only find references to a subgraph in contexts (like `subgraph = ${subgraph}`) no occurrence of `.subgraph` - Am I missing something?!

Reply:
@bwaidelich you are right, we might be able to remove this :-)

Reply:
Awesome!